### PR TITLE
Check for DefaultWebApp setting before prompting

### PR DESCRIPTION
**Pull request by nturinski**
Fixes https://github.com/microsoft/vscode-azureappservice/issues/1077

The previous check didn't work because `deployWithConfigs` was set to false when there wasn't a default web app.  This didn't account for the `none` behavior.

I wanted to change deployWithConfigs to include a `none` type, but `getDefaultWebApp` doesn't get called if the user entered via the context menu (since we already have a node) so I have to check the workspace setting instead.

### CODE REVIEW

File: src/commands/deploy/deploy.ts [hunk 1]
```diff
@@ -158,7 +158,8 @@ export async function deploy(context: IActionContext, confirmDeployment: boolean
         deployContext.telemetry.properties.cancelStep = '';
     }
 
-    if (!deployContext.deployedWithConfigs) {
+    const defaultWebAppToDeploySetting: string | undefined = getWorkspaceSetting(constants.configurationSettings.defaultWebAppToDeploy, deployFsPath);
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/deploy.ts]: > I wanted to change deployWithConfigs to include a none type, but getDefaultWebApp doesn't get called if the user entered via the context menu (since we already have a node) so I have to check the workspace setting instead.

We usually have a separate setting like `appService.showDefaultWebAppPrompt` that is checked regardless of the entry point. Instead, the `none` value essentially mashes the `defaultWebAppToDeploy` setting with `showDefaultWebAppPrompt` which is why you ran into that problem. That being said I wouldn't want to change this without discussing backwards compatibility.

File: src/commands/getDefaultWebAppToDeploy.ts [hunk 2]
```diff
@@ -11,8 +11,9 @@ import { IDeployWizardContext } from "./deploy/IDeployWizardContext";
 
 export async function getDefaultWebAppToDeploy(context: IDeployWizardContext): Promise<WebAppTreeItem | undefined> {
```
Comment (EricJizbaMSFT) [inline src/commands/getDefaultWebAppToDeploy.ts]: This file should be in the `deploy` folder

File: src/commands/getDefaultWebAppToDeploy.ts [hunk 3]
```diff
@@ -11,8 +11,9 @@ import { IDeployWizardContext } from "./deploy/IDeployWizardContext";
...
     const defaultWebAppId: string | undefined = getWorkspaceSetting(configurationSettings.defaultWebAppToDeploy, context.workspace.uri.fsPath);
-
-    if (defaultWebAppId && defaultWebAppId !== none) {
+    if (defaultWebAppId === none) {
+        context.telemetry.properties.deployedWithConfigs = 'none';
+    } else if (defaultWebAppId) {
         const defaultWebApp: WebAppTreeItem | undefined = await ext.tree.findTreeItem(defaultWebAppId, context); // resolves to undefined if app can't be found
         if (defaultWebApp) {
             context.deployedWithConfigs = true;
```
Comment (EricJizbaMSFT) [inline src/commands/getDefaultWebAppToDeploy.ts]: The name `deployedWithConfigs` doesn't make sense to me. Which configs is it talking about? I would guess defaultWebAppToDeploy _and_ deploySubpath based on the name, but as far as I can tell only `defaultWebAppToDeploy` matters for this. As a possible alternative, we could refactor to have a `webAppSource` property that allows the values `setting`, `rightClick` or `nodePicker`

File: src/commands/deploy/deploy.ts [hunk 4]
```diff
@@ -158,7 +158,8 @@ export async function deploy(context: IActionContext, confirmDeployment: boolean
         deployContext.telemetry.properties.cancelStep = '';
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/deploy.ts]: I think another root cause is disabling `max-func-body-length cyclomatic-complexity` in `deploy` because it makes it hard to follow all of the possible code-paths. Would like to see that fixed sometime soon, but not in this PR

File: src/commands/deploy/deploy.ts (hunk 1 above)
Comment (nturinski) [reply src/commands/deploy/deploy.ts]: I could write migration code that checks for the `none` value in showDefaultWebApp prompt and then sets `appService.showDefaultWebAppPrompt`.  That seems simple enough, but is an ugly solution that would have to remain in the codebase _FOREVER_ 😱

File: src/commands/getDefaultWebAppToDeploy.ts (hunk 3 above)
Comment (nturinski) [reply src/commands/getDefaultWebAppToDeploy.ts]: Yeah, agreed. I think back when I made that property, we never talked about using `deploySubpath` settings.  I think I'd prefer a `webAppSource` property as you describe as the subDeployPath stuff should be handled by the `appservice` package now.

File: src/commands/deploy/deploy.ts (hunk 4 above)
Comment (nturinski) [reply src/commands/deploy/deploy.ts]: I'll assign this to @wwlorey 😁 

jk, but yes, I agree.  I'll start refactoring this after the release.  This code is getting to be a mess: https://github.com/microsoft/vscode-azureappservice/issues/1112

File: src/commands/deploy/deploy.ts [hunk 5]
```diff
@@ -21,23 +21,26 @@ import { getRandomHexString } from "../../utils/randomUtils";
...
 import { getWorkspaceSetting, updateWorkspaceSetting } from '../../vsCodeConfig/settings';
-import { getDefaultWebAppToDeploy } from '../getDefaultWebAppToDeploy';
 import { startStreamingLogs } from '../startStreamingLogs';
-import { IDeployWizardContext } from './IDeployWizardContext';
+import { getDefaultWebAppToDeploy } from './getDefaultWebAppToDeploy';
+import { IDeployWizardContext, WebAppSource } from './IDeployWizardContext';
 
 // tslint:disable-next-line:max-func-body-length cyclomatic-complexity
 export async function deploy(context: IActionContext, confirmDeployment: boolean, target?: vscode.Uri | SiteTreeItem | undefined): Promise<void> {
+    context.telemetry.properties.deployedWithDefaultApp = 'false';
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/deploy.ts]: Can we use the same `webAppSource` property for telemetry? It would be more consistent, easier to write a query for in telemetry, and I think less code since you'd only have to set it once after `webAppSource` has been set

File: src/commands/deploy/deploy.ts [hunk 6]
```diff
@@ -21,23 +21,26 @@ import { getRandomHexString } from "../../utils/randomUtils";
...
-    context.telemetry.properties.deployedWithConfigs = 'false';
+    let node: SiteTreeItem | undefined;
     let deployFsPath: string | undefined;
+    let webAppSource: WebAppSource;
 
     if (target instanceof vscode.Uri) {
-        deployFsPath = target.fsPath;
         context.telemetry.properties.deploymentEntryPoint = 'fileExplorerContextMenu';
+        deployFsPath = target.fsPath;
+        webAppSource = WebAppSource.nodePicker;
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/deploy.ts]: We don't actually know if this will be `nodePicker` until later on. I would prefer to delay setting this value until we actually know what it will be. So instead of this:
```
node = await getDefaultWebAppToDeploy(deployContext) || <SiteTreeItem>await ext.tree.showTreeItemPicker(WebAppTreeItem.contextValue, context);
```

It might look like this:
```
node = await getDefaultWebAppToDeploy(deployContext);
if (node) {
    webAppSource = WebAppSource.tree;
} else {
    webAppSource = WebAppSource.nodePicker;
    node = <SiteTreeItem>await ext.tree.showTreeItemPicker(WebAppTreeItem.contextValue, context);
}
```

File: src/commands/deploy/deploy.ts (hunk 4 above)
Comment (EricJizbaMSFT) [reply src/commands/deploy/deploy.ts]: As an example, I think you could make `getDefaultWebAppToDeploy` broader to just be `getWebAppToDeploy` similar to how I had `getDeployFsPath`. In other words it takes in whatever the target is and returns a web app after checking all the possible settings, node picker, target, etc.

File: src/commands/deploy/deploy.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply src/commands/deploy/deploy.ts]: It wouldn't have to be there _forever_. We could deprecate the old setting and remove it eventually

File: src/commands/deploy/deploy.ts (hunk 6 above)
Comment (nturinski) [reply src/commands/deploy/deploy.ts]: webAppSource gets overwritten in getDefaultWebApp to `WebAppSource.setting` if we used the setting, though this might not be good coding practice 🤔

File: src/commands/deploy/deploy.ts (hunk 5 above)
Comment (nturinski) [reply src/commands/deploy/deploy.ts]: Well, webAppSource will always be nodePicker | setting | treeView, but deployedWithDefaultApp is being used to track whether or not people are using the defaultWebAppSetting including if they are using the `none` setting.

Comment (nturinski) [reply src/commands/deploy/deploy.ts]: That being said, we track if people click the "Don't show again" button, so I don't think we necessarily have to track the 'none' setting.
